apps/bot/src/services/twitch-eventsub/index.ts

```diff
@@ -0,0 +1,131 @@
+import { logger } from "utils/logger";
+import { type MessageEvent, WebSocket } from "ws";
+import { SEventSubGeneralMessage, SEventSubMetadata } from "./types";
+import { handleEvents } from "./events";
+
+export class TwitchEventSub {
+  private ws: WebSocket;
+
+  private static _instance: TwitchEventSub | undefined;
+
+  private constructor(private url: string) {
+    this.ws = new WebSocket(url);
+  }
+
+  public static instance(url: string) {
+    if (!TwitchEventSub._instance) {
+      TwitchEventSub._instance = new TwitchEventSub(url);
+      return TwitchEventSub._instance;
+    }
+
+    return TwitchEventSub._instance;
+  }
+
+  public async connect(): Promise<void> {
+    await new Promise((resolve) => {
+      try {
+        this.ws.addEventListener("open", () => {
+          console.log("connection is open");
+        });
+        this.ws.addEventListener("message", (r) => this.onMessage(r));
+      } catch (_) {
+        resolve(undefined);
+      }
+    });
+  }
+
+  private handlePingMessage(msg: string): void {
+    if (msg.startsWith("PING")) {
+      const pong = `PONG${msg.substring("PING".length)}`;
+      this.ws.send(pong);
+    }
+  }
+
+  private keepaliveTimeout = 10;
+  private lastevent = 0;
+
+  private async reconnect(url?: string) {
+    logger.warning("[EventSub] Triggering reconnect");
+    this.ws.removeAllListeners();
+    this.ws.close();
+
+    const lurl = url ?? this.url;
+    this.ws = new WebSocket(lurl);
+
+    await this.connect();
+  }
+
+  private startPingCheck() {
+    setInterval(() => {
+      this.lastevent = this.lastevent + 1;
+      console.log(this.lastevent, this.keepaliveTimeout);
+      if (this.lastevent > this.keepaliveTimeout) {
+        this.reconnect();
+      }
+    }, 1000);
+  }
```

Comment:
this method `startPingCheck`, a little bit of food for thought...
without a limit for the number of times it tries to reconnect, it could create some "interesting" small bugs in the long run, but we are already using JS on the server, what can't go wrong xD